core/include/drivers/pl022_spi.h

```diff
@@ -34,17 +34,26 @@
 
 #define PL022_REG_SIZE	0x1000
 
+struct pl022_cs_gpio_data {
+	struct gpio_chip	*chip;
+	unsigned int		pin_num; /* chip select */
+};
+
+union pl022_cs_data {
+	struct pl022_cs_gpio_data	gpio_data;
+	void				(*cs_cb)(enum gpio_level value);
+};
+
 struct pl022_data {
+	union pl022_cs_data	cs_data;
 	struct spi_chip		chip;
-	struct gpio_chip	*gpio;
-	vaddr_t		base;
-	vaddr_t		cs_gpio_base; /* chip select */
-	unsigned int	clk_hz;
-	unsigned int	speed_hz;
-	unsigned int	cs_gpio_pin; /* chip select */
-	unsigned int	mode;
-	unsigned int	data_size_bits;
-	bool		loopback;
+	vaddr_t			base;
+	unsigned int		clk_hz;
+	unsigned int		speed_hz;
+	unsigned int		mode;
+	unsigned int		data_size_bits;
+	bool			b_cs_callback;
+	bool			b_loopback;
```

Comment:
I'd rather have `b_cs_callback` and `loopback` (because `cs_callback` would be ambiguous, could be the pointer itself, no such issue with `loopback`).